Start central serialization methods

lkastner opened the pull request:
Since I started this PR it has been refactored after feedback from various people. Now the serialization is done by creating functions `save_intern` and `load_intern` for the different types we want to serialize and these functions pass on structs `SerializerState` and `DeserializerState` for recording the state, to allow backrefs and recording the context.

Obviously I cannot create these methods for everything in Oscar all by myself, so I hope that this is acceptable as a blueprint/draft.

lkastner committed:
Serialization: First steps with state

- Introduce (Des|Se)rializerState for avoiding duplication and for
  storing context
- Refactor previous simple interface for polymake serialization
- Use JSON, but other formats should be possible

fingolfin commented on src/PolyhedralGeometry/Serialization.jl:
```diff
@@ -1,170 +1,131 @@
 ##############################################################################
 """
-    save_polyhedron(Cone, String)
+    save(Cone, String)
```
This does not match the code below, nor does it fit into the conventions for docstrings. Should be like this (also applies elsewhere)
```suggestion
    save(P::Polyhedron, filename::String)
```

lkastner replied:
Thanks

fingolfin commented on src/PolyhedralGeometry/Serialization.jl:
```diff
@@ -1,170 +1,131 @@
-
-Load a cone stored in JSON format, given the filename as input.
-"""
-function load_cone(filename::String)
```
An alternative approach would be to replace `load_cone(str)` by `load(Cone, str)` (where `Cone` literally is the type, passed in as an argument). I.e.
```
function load(::Type{Cone}, filename::String)
```

lkastner replied:
Well, but then the user needs to provide what they think this object is anyway. I can do this though, if this is preferred.

fingolfin replied:
I can't say what is "preferred" as I still don't know what the big plan is for using these things. All I am saying is: if one is willing to write `load_cone(str)` then one should also be willing to accept `load(Cone, str)`, but that one has the advantage of being easier to use in generic code.

fieker commented:
On Wed, Jan 19, 2022 at 03:09:41PM -0800, Lars Kastner wrote:
> The existing serialization methods are refactored into two central methods `save` and `load` that recognize the datatype from the JSON file.
> - Add JSON as dependency
> - Currently I only do the polyhedral objects, but I clearly mark the places where one would place a hook for loading other software.
> - The `save` side is easy, as one only needs to define a `save` function for every object.
> 
> I am aware of that the return type of `load` will be very ambiguous, I have raised this issue myself and this was the reason for the previous structure of the code. I have been told though, that a central `save` and `load` are desired.

It is a start and I am glad. However, I still would like to understand/
diskuss how this can be made to work in a "more complicated" setting,
i.e. where parent objects are involved. The current syntax, I think,
cannot be made to work for that. In order to have a stackable system
that allows more complicated objects a context/ dictionary/ ... has to
be included somewhere I think.

Consider a polynomial over some number field, start with a single
polynomial.
To serialize:
 - the number field needs to be involved
 - then I can serialize the degree
 - then a number of elements of the field
   A number field element is potentially a polynomial over a smaller
   field, thus, ...
   In particular, when reading the polynomial in, all coefficients need
   to be in the identical field, hence the field has to be part of the
   reading interface.

As a model I can see s.th. like
 - create a serialization context (like an ID-Dict)
 - for every object in this call
   see if it is already serialized, then send a reference (using the
   dict)
   else send the object and store the info in the dict.
For this to work, every serialization/ deserialization function needs to
accept the context/ dict as an argument... this is why we should discuss
this BEFORE writing many such functions...

> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/oscar-system/Oscar.jl/pull/1000
> 
> -- Commit Summary --
> 
>   * Start refactoring serialization to unified save+load
> 
> -- File Changes --
> 
>     M Project.toml (2)
>     M src/PolyhedralGeometry/Serialization.jl (172)
>     M src/PolyhedralGeometry/main.jl (12)
>     M test/PolyhedralGeometry/Serialization.jl (20)
> 
> -- Patch Links --
> 
> https://github.com/oscar-system/Oscar.jl/pull/1000.patch
> https://github.com/oscar-system/Oscar.jl/pull/1000.diff
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/1000
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***&gt;

fingolfin commented on src/PolyhedralGeometry/Serialization.jl:
```diff
@@ -1,170 +1,211 @@
-      throw(ArgumentError("Loaded object is not of polymake type Cone, it has type " * typename))
-   end
-   return Cone(bigobject)
+function load(as::Type{Cone}, filename::String)
```
```suggestion
function load(::Type{Cone}, filename::String)
```

lkastner replied:
Thanks, makes sense, wasn't clear to me.

fingolfin commented on docs/src/PolyhedralGeometry/cones.md:
```diff
@@ -43,19 +43,14 @@ Objects of type `Cone` can be saved to a file and loaded from a file in the
-save_cone(C, "C.cone")
-CC = load_cone("C.cone")
+save(C, "C.cone")
+CC = load("C.cone")
```
Should this perhaps be
```suggestion
CC = load(Cone, "C.cone")
```
?

lkastner replied:
So the question is, whether we can have both, i.e. `load(filename)` and `load(Type, filename)`. The latter for places where type inference matters for performance reasons, the former for convenience, i.e. inexperienced users, we don't have to explain to people how to figure out what is in the file or how to tell their colleagues. I now added both, and I think it doesn't hurt to have `load(filename)` for now. If it becomes too unmanageable we can always eliminate it. Is there some strong reason against that I don't see?

fieker commented:
Will there be a version with a stream or file passed in? To allow multiple
objects in the same file? This would be necessary for building and
serialising complex structures

On Sat, 22 Jan 2022, 00:55 Max Horn, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/PolyhedralGeometry/Serialization.jl
> <https://github.com/oscar-system/Oscar.jl/pull/1000#discussion_r790064405>
> :
>
> >
>
>  Load a cone stored in JSON format, given the filename as input.
>
>  """
>
> -function load_cone(filename::String)
>
> -   bigobject = Polymake.load_bigobject(filename)
>
> -   typename = Polymake.type_name(bigobject)
>
> -   if typename != "Cone<Rational>"
>
> -      throw(ArgumentError("Loaded object is not of polymake type Cone, it has type " * typename))
>
> -   end
>
> -   return Cone(bigobject)
>
> +function load(as::Type{Cone}, filename::String)
>
>
> ⬇️ Suggested change
>
> -function load(as::Type{Cone}, filename::String)
>
> +function load(::Type{Cone}, filename::String)
>
>
> ------------------------------
>
> In docs/src/PolyhedralGeometry/cones.md
> <https://github.com/oscar-system/Oscar.jl/pull/1000#discussion_r790064635>
> :
>
> > @@ -43,19 +43,14 @@ Objects of type `Cone` can be saved to a file and loaded from a file in the
>
>  following way:
>
>  ***@***.*** oscar
>
>  C = positive_hull([1 0; 0 1])
>
> -save_cone(C, "C.cone")
>
> -CC = load_cone("C.cone")
>
> +save(C, "C.cone")
>
> +CC = load("C.cone")
>
>
> Should this perhaps be
> ⬇️ Suggested change
>
> -CC = load("C.cone")
>
> +CC = load(Cone, "C.cone")
>
>
> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oscar-system/Oscar.jl/pull/1000#pullrequestreview-860124220>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA36CV4UMA6MSW5VW7MFSZ3UXHXAXANCNFSM5MLHKXIA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

lkastner commented:
> Will there be a version with a stream or file passed in? To allow multiple objects in the same file? This would be necessary for building and serialising complex structures

Yes, this is the long-term structure I have in mind. Currently it detects that the objects passed in come from polymake via the namespace parameter. So a new structure can either leave out this parameter or introduce its own namespace. We would probably have a oscar namespace and then use another parameter to detect the type, then it should be possible to recurse.

Nevertheless for now I just wanted to enable reading and writing for the polyhedral objects via Polymake.jl but I tried to keep everything as flexible as possible by modularizing with these future developments in mind. However there are still many things to discuss, as you mentioned.

micjoswig commented:
> However, I still would like to understand/ diskuss how this can be made to work in a "more complicated" setting, i.e. where parent objects are involved.  ...

This was discussed (in some detail) at the recent polymake workshop.  The scenario we were considering is (still) "convex hulls over number fields"; so that should be general enough.  Here is the rough sketch.

(1) OSCAR types without polymake equivalent will live in a separate name space. Objects from several name spaces may live in the same file together, but more often there is one top level object from one name space.  In the more distant future it is conceivable to have further name spaces (beyond polymake and OSCAR).

(2) We will allow for pointers.  These may come in two flavors: relative via paths in the JSON tree (this includes, e.g., parents) and absolute (via URI).

Ewgenij created a sketch; will be discussed, then implemented.  The goal is to make the existing data base function accessible to all objects with such a serialization.

lkastner committed:
PolyhedralGeometry: Do not copy underlying polyhedron for LP

lkastner committed:
Serialization: Simplify recording version with builtin variables

lkastner committed:
Serialization: Sparsely document save and load

lkastner committed:
Serialization: First shot at serializing vectors of our objects

lkastner committed:
Serialization: Use UUIDs instead of simple IDs

lkastner committed:
Serialization: Draft for serializing GF(p)

lkastner committed:
Serialization: Refactor vector and files

- Solve Vector storing more elegantly
- Better file structure

lkastner commented:
@fingolfin @benlorenz Do you think this is ok for now? I tried to follow along the lines of @fingolfin 's writeup (Thanks a lot btw :wink: ). I don't want to make this PR any larger, so if you agree, I would merge and then work on further parts in separate PRs.

lkastner committed:
Serialization: Deprecate old serialization functions

lkastner committed:
Serialization: Add test with actual file

fingolfin commented on src/Serialization/Fields.jl:
```diff
@@ -0,0 +1,50 @@
+end
+
+function load_intern(s::DeserializerState, F::Type{Nemo.GaloisField}, dict::Dict)
+    return F(UInt64(dict[:characteristic]))
```
Note that `F === Nemo.GaloisField` here, so you could rewrite this code as saying
```suggestion
function load_intern(s::DeserializerState, ::Type{Nemo.GaloisField}, dict::Dict)
    return Nemo.GaloisField(UInt64(dict[:characteristic]))
```
Similar elsewhere.

fingolfin reviewed (approved):
Needs a rebase. Do we want/need to discuss this further, @fieker @lkastner ? Or merge it and discuss later? Either way is fine by me.

fingolfin commented:
(Before merging, a conflict needs to be resolved)

lkastner commented:
Benjamin suggested to prefix things that come from (or go to) polymake with a symbol `:polymake`. Another idea was to further automate serialization of our objects by just looping over their properties, but the latter part seems to fail (for now). I will implement the first suggestion and then merge.

lkastner committed:
Serialization: Scalar params from polyhedral geometry

- Adapt serialization of polyhedral objects to design change.
- Make serialization more flexible overall, by allowing serialization of
  types unknown to the typeMap.

lkastner committed:
Serialization: Generic serialization prototype

- Added a prototype for serializing a bit more general OSCAR objects.
- Polyhedral objects are now serialized via the generic variant.

lkastner committed:
Update src/Serialization/Fields.jl

Co-authored-by: Max Horn <max@quendi.de>

lkastner committed:
Serialization: Do not record Julia version

lkastner committed:
Serialization: Optionally skip check for namespace, necessary for mixed containers

lkastner committed:
Serialization: Allow weird vectors
